Re-enable Op codes for May 15 hardfork

ptschip opened the pull request.

deadalnix committed:
Add missing test cases in script_tests.cpp

Summary: Some tests case were present in script_tests.json but missing from script_tests.cpp . This is fixing that error.

Test Plan:
  make check

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1197

Danconnolly committed:
Prepare for re-enabled opcodes for the May 2018 protocol upgrade

Summary:
  - Rename OP_SUBSTR to OP_SPLIT
  - Add OP_BIN2NUM & OP_NUM2BIN, replacing OP_LEFT and OP_RIGHT
  - tests in script_tests.json were updated to use the renamed opcodes
  - Also added FIRST_UNDEFINED_OP_VALUE, the first value of undefined opcodes - OP_NOP10+1 was being used for this purpose which is confusing

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>
Co-authored-by: Daniel Connolly <daniel@dconnolly.com>

Test Plan:
make check
test/functional/test_runner.py --extended

Reviewers: #bitcoin_abc, O1 Bitcoin ABC, schancel, movrcx

Reviewed By: #bitcoin_abc, O1 Bitcoin ABC, schancel

Subscribers: schancel, teamcity, danconnolly, deadalnix

Maniphest Tasks: T303

Differential Revision: https://reviews.bitcoinabc.org/D1141

deadalnix committed:
Add a flag to gate opcodes activation for the new HF

Summary: This per se does nothing except adding a flag to gate opcodes and the infrastructure to use this flag in tests.

Test Plan: Test involving the flag have been added for all disabled opcodes.

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1208

deadalnix committed:
Explode disabled opcode if into a switch

Summary: This ensures that they can be manipulated one by one more easily.

Test Plan:
  make check

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1207

schancel committed:
Add helper function for disabled opcodes

Summary:
Add `IsOpcodeDisabled` helper function to see if an opcode
is supposed to be disabled or not

Test Plan: `make check`

Reviewers: deadalnix, jasonbcox, movrcx, matiu, #bitcoin_abc

Reviewed By: jasonbcox, #bitcoin_abc

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1209

schancel committed:
Pull minimal check out of CScriptNum constr. into IsMinimalArray

Summary:
Move the check for minimal encoded arrays out of CScriptNum's
constructor and into a free function.  This enables it to be used
elsewhere, and avoid hitting exceptions.

Test Plan: `make check`

Reviewers: movrcx, jasonbcox, matiu, deadalnix, #bitcoin_abc

Reviewed By: jasonbcox, #bitcoin_abc

Subscribers: shadders, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1214

jasonbcox committed:
Add function to trim leading array zeros maintaining MSB.

Summary:
Add a function `MinimalizeBigEndianArray` to trim a byte array
of leading zeros, while maintaining any MSB for negation. This
will get used in BIN2NUM in a later patch.

 Co-authored-by: Shammah Chancellor <shammah.chancellor@gmail.com>

Test Plan: `make check`

Reviewers: shadders, movrcx, #bitcoin_abc, schancel, deadalnix

Reviewed By: #bitcoin_abc, schancel

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1215

deadalnix committed:
Simplify IsOpcodeDisabled

Summary:
Cyclomatic complexity goes from 27 to 17.

Also remove inline qualifier that is completely useless here.

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1217

deadalnix committed:
Add support for AND, OR and XOR opcodes

Summary:
These opcodes are gated by a flag and compute what you'd expect for
binary ops. Added cases to the json test cases. In addition, added a
new unit tests that check if the opcodes behave in the expected way
in a large variety of situations.

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>
Co-authored-by: Daniel Connolly <daniel@dconnolly.com>
Co-authored-by: Shammah Chancellor <shammah.chancellor@bitcoinabc.org>

Test Plan: `make check`

Reviewers: #bitcoin_abc, jasonbcox, schancel

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Maniphest Tasks: T304

Differential Revision: https://reviews.bitcoinabc.org/D1211

deadalnix committed:
Ensure that CScriptNum is self contained.

Summary: CScriptNum is somewhat of a mess, so ensuring the mess is contained is good.

Test Plan:
  make check

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1218

deadalnix committed:
Refactor MinimalizeBigEndianArray t be part of CScriptNum

Summary: Also make the logic LE, like everything related to numbers

Test Plan: Move test case over to scriptnum_tests.cpp . Add check against CScriptNum::IsMinimallyEncoded to make sure both are consistent.

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1219

deadalnix committed:
Add support for BIN2NUM opcode

Summary: As per title. As the whole number system is LE, I went went for LE encoding. This doesn't seems like a great idea to fubar the scrpipt system even more than it already is by introducing mixed endianess in it.

Test Plan: Added various test cases.

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, schancel, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1220

deadalnix committed:
Refactor opcode test to reduce redundancy

Summary: This will ensure that newly created tests will test all the flags painlessly.

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox

Differential Revision: https://reviews.bitcoinabc.org/D1221

deadalnix committed:
Implement support for NUM2BIN

Summary: As per title.

Test Plan: Ass json test cases and unit tests.

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: jasonbcox, teamcity

Maniphest Tasks: T310

Differential Revision: https://reviews.bitcoinabc.org/D1222

jasonbcox committed:
Refactored existing monolith_opcodes tests

Summary:
These changes are needed in order to merge D1212 and D1097

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>

Test Plan: make check

Reviewers: deadalnix, schancel, #bitcoin_abc

Reviewed By: deadalnix, #bitcoin_abc

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1223

jmprcx committed:
Add support for SPLIT opcodes

Summary:
Add handling for SPLIT opcodes when the monolith script flag is enabled.

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>

Test Plan: `make check`

Reviewers: #bitcoin_abc, O1 Bitcoin ABC, movrcx, jasonbcox

Reviewed By: #bitcoin_abc, O1 Bitcoin ABC, jasonbcox

Subscribers: jasonbcox, shadders, danconnolly, teamcity, macros

Maniphest Tasks: T308

Differential Revision: https://reviews.bitcoinabc.org/D1228

jmprcx committed:
Add support for CAT

Summary:
Add handling for CAT opcode when the monolith script flag is enabled.

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>

Test Plan: `make check`

Reviewers: #bitcoin_abc, O1 Bitcoin ABC, movrcx, jasonbcox

Reviewed By: #bitcoin_abc, O1 Bitcoin ABC, jasonbcox

Subscribers: jasonbcox, shadders, danconnolly, teamcity, macros

Maniphest Tasks: T307

Differential Revision: https://reviews.bitcoinabc.org/D1227

Danconnolly committed:
Set Monolith opcode enable script flag if monolith enabled

Summary:
Sets the SCRIPT_ENABLE_OPCODE_MONOLITH flag if monolith has been enabled

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>
Co-authored-by: Daniel Connolly <daniel@dconnolly.com>

Test Plan:
  make check

Added an integration test for activation.

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Maniphest Tasks: T299

Differential Revision: https://reviews.bitcoinabc.org/D1231

jasonbcox committed:
Add support for DIV and MOD opcodes

Summary:
Added implementation for OP_DIV and OP_MOD with minimal tests

Depends on D1223

Co-authored-by: Joshua Yabut <yabut.joshua@gmail.com>
Co-authored-by: Marcos Mayorga <mm@mm-studios.com>
Co-authored-by: Daniel Connolly <daniel@dconnolly.com>
Co-authored-by: Shammah Chancellor <shammah.chancellor@bitcoinabc.org>

Test Plan: make check

Reviewers: schancel, #bitcoin_abc, deadalnix

Reviewed By: #bitcoin_abc, deadalnix

Subscribers: Mengerian, movrcx, deadalnix, macros, teamcity

Maniphest Tasks: T306, T305

Differential Revision: https://reviews.bitcoinabc.org/D1212

ptschip committed:
Cleanup scriptflags in unit tests

Use only one set of script flags in scriptflags.cpp.

Add MONOLITH_OPCODES and REPLAY_PROTECTION in mapFlagNames.

add new header file src/test/scriptflags.h

ptschip committed:
Fix abc-monolith-acitivation.py test failures

  Add may2018activationTime to regtest chainparams.
  This was missing and is needed for python tests to work.

  Fix script timing issue where we can't be sure of ther return
  code when checking the block. We may get a signature failure
  or a checkinputs failure. All we can do is be certain that we
  do get a failure.

joshuayabut commented:
I've built this locally and tested it. Looks good to me. +1

sickpig commented on qa/rpc-tests/abc-monolith-activation.py:
```diff
@@ -49,11 +51,12 @@ def create_and_tx(self, count):
 
     def run_test(self):
         self.test = TestManager(self, self.options.tmpdir)
+        print ("num nodes is " + str(len(self.nodes)))
```
I would remove this print statement before merging

ptschip replied:
done

sickpig commented on qa/rpc-tests/abc-monolith-activation.py:
```diff
@@ -91,15 +94,15 @@ def spend_and():
 
         # Check that large opreturn are not accepted yet.
-        self.log.info("Try to use the monolith opcodes before activation")
+        print("Try to use the monolith opcodes before activation")
```
ditto

ptschip replied:
switched to loggin.info()

sickpig commented on qa/rpc-tests/abc-monolith-activation.py:
```diff
@@ -91,15 +94,15 @@ def spend_and():
 
         # Push MTP forward just before activation.
-        self.log.info("Pushing MTP just before the activation and check again")
+        print("Pushing MTP just before the activation and check again")
```
ditto

ptschip replied:
switched to loggin.info()

sickpig commented on qa/rpc-tests/abc-monolith-activation.py:
```diff
@@ -145,24 +148,31 @@ def add_tx(block, tx):
+        yield rejected(b)
 
-        self.log.info("Activates the new opcodes")
+        print("Activates the new opcodes")
```
ditto

ptschip replied:
switched to loggin.info()

sickpig commented:
great work @ptschip, will review later today.

ptschip committed:
remove unnecessary print statement from python test

ptschip commented:
added -monolithactivationtime to allowed_args.  That should be the final commit other than anything else that shows up from review.

gandrewstone commented:
We have been using this flag to trigger the fork:  mining.forkMay2018Time
this is redundant with "-monolithactivationtime"

Let's go with mining.forkMay2018Time since it is descriptive.
"monolith" is maybe an ABC code word?  It has no descriptive meaning, but kind of implies meaning (unlike something obviously codeword like: "BlueHawk"), so I'd like it to be removed from function calls, and file names.

Fork enabling function calls already exist but should use the date May2018 in the name.
File names could use subject based names like "opcode".

ptschip committed:
Use may152018 rather than monolith to describe files and variables

related to the hard fork.

ptschip commented:
@gandrewstone ok I changed monolith to may152018 for variables and filenames/testnames etc.

Also dropped the -monolithactivationtime arg and sticking with mining.forkMay15 ...

ptschip committed:
Move activation for EB, MG and data carrier size to UpdateTip

Because PV can potentially connect several blocks together at one
time it is more accurate to check here just after we've updated
the chain tip.  Also this means that when a block is disconnected
the check also happens.

ptschip committed:
Make both May fork tests have similar names to avoid confusion.

gandrewstone committed:
merge changes from dev

gandrewstone committed:
remove duplicate fork activation logic

ptschip committed:
Merge pull request #11 from gandrewstone/ptschip-dev_opcodes

remove redundant fork logic
